torch/csrc/autograd/custom_function.h

```diff
@@ -41,11 +56,37 @@ struct CFunction {
   static variable_list apply(Args&&... args);
 };
 
+
+// Context to save information during forward that can be accessed in backward
 struct AutogradContext {
-  std::vector<torch::autograd::SavedVariable> saved_variables;
+  // Can be used to save non-variable data for backward()
+  std::unordered_map<std::string, at::IValue> save;
+
+  // Saves the list of variables for a future call to backward(). This
+  // should be called at most once from inside of forward().
+  void save_for_backward(const variable_list &to_save);
+  // Marks variables in the list as modified in an in-place opertaion. This
+  // should be called at most once from inside of forward() and all arguments
+  // should be inputs.
+  void mark_dirty(const variable_list &inputs);
+  // Marks outputs in the list as not requiring gradients. This should be called
+  // at most once from inside of forward() and all arguments should be outputs.
+  void mark_non_differentiable(const variable_list &outputs);
+  void clear_saved();
 
+  // Get the list of variables that were saved in forward using
+  // save_for_backward(). Before returning them to the user, a check is made to
+  // ensure that they were not modified by any in-place operations.
+  variable_list get_saved_variables() const;
+  const std::unordered_set<at::TensorImpl*>& get_dirty() const;
+  const std::unordered_set<at::TensorImpl*>& get_non_differentiable() const;
+
+private:
   std::unordered_set<at::TensorImpl*> non_differentiable;
   std::unordered_set<at::TensorImpl*> dirty_inputs;
+  std::vector<torch::autograd::SavedVariable> saved_variables;
+
+  std::shared_ptr<Function> grad_fn;
```

Comment:
Is it ok in cpp to have a shared_ptr that contains an object that contains this shared pointer?
The Function owns the context here and then context seems to own the Function as well. Or I got lost somewhere?

Reply:
Yeah, this is probably causing a refcycle.